logins-sql/src/schema.rs

```diff
@@ -0,0 +1,304 @@
+            server_modified INTEGER NOT NULL,
+            is_overridden   TINYINT NOT NULL DEFAULT 0
+        )",
+        common_sql = COMMON_SQL
+    );
+
+    static ref SET_VERSION_SQL: String = format!(
+        "PRAGMA user_version = {version}",
+        version = VERSION
+    );
+}
+
+const CREATE_META_TABLE_SQL: &'static str = "
+    CREATE TABLE IF NOT EXISTS loginsSyncMeta (
+        key TEXT PRIMARY KEY,
+        value NOT NULL
+    )
+";
+
+const CREATE_OVERRIDE_HOSTNAME_INDEX_SQL: &'static str = "
+    CREATE INDEX IF NOT EXISTS idx_loginsM_is_overridden_hostname
+    ON loginsM (is_overridden, hostname)
+";
+
+const CREATE_DELETED_HOSTNAME_INDEX_SQL: &'static str = "
+    CREATE INDEX IF NOT EXISTS idx_loginsL_is_deleted_hostname
+    ON loginsL (is_deleted, hostname)
+";
+
+// As noted above, we use these when updating from schema v3 (firefox-ios's
+// last schema) to convert from microsecond timestamps to milliseconds.
+const UPDATE_LOCAL_TIMESTAMPS_TO_MILLIS_SQL: &'static str = "
+    UPDATE loginsL
+    SET timeCreated = timeCreated / 1000,
+        timeLastUsed = timeLastUsed / 1000,
+        timePasswordChanged = timePasswordChanged / 1000
+";
+
+const UPDATE_MIRROR_TIMESTAMPS_TO_MILLIS_SQL: &'static str = "
+    UPDATE loginsM
+    SET timeCreated = timeCreated / 1000,
+        timeLastUsed = timeLastUsed / 1000,
+        timePasswordChanged = timePasswordChanged / 1000
+";
+
+pub(crate) static LAST_SYNC_META_KEY:    &'static str = "last_sync_time";
+pub(crate) static GLOBAL_STATE_META_KEY: &'static str = "global_state";
+
+pub(crate) fn init(db: &db::LoginDb) -> Result<()> {
+    let user_version = db.query_one::<i64>("PRAGMA user_version")?;
+    if user_version == 0 {
+        // This logic is largely taken from firefox-ios. AFAICT at some point
+        // they went from having schema versions tracked using a table named
+        // `tableList` to using `PRAGMA user_version`. This leads to the
+        // following logic:
+        //
+        // - If `tableList` exists, we're hopelessly far in the past, drop any
+        //   tables we have (to ensure we avoid name collisions/stale data) and
+        //   recreate. (This is captured by the `upgrade` case where from == 0)
+        //
+        // - If `tableList` doesn't exist and `PRAGMA user_version` is 0, it's
+        //   the first time through, just create the new tables.
+        //
+        // - Otherwise, it's a normal schema upgrade from an earlier
+        //   `PRAGMA user_version`.
+        let table_list_exists = db.query_one::<i64>(
+            "SELECT count(*) FROM sqlite_master WHERE type = 'table' AND name = 'tableList'"
+        )? != 0;
+
+        if table_list_exists {
+            drop(db)?;
+        }
+        return create(db);
+    }
+    if user_version != VERSION {
+        if user_version < VERSION {
+            upgrade(db, user_version)?;
+        } else {
+            warn!("Loaded future schema version {} (we only understand version {}). \
```

Comment:
Should we downgrade the schema version, too, so that any rows changed by us get migrated when opened with a newer client? Or set `PRAGMA query_only = 1`, so that we can't write? (Or just not worth the trouble? :smile:)

Reply:
Downgrading the schema would cause the next updated consumer to try to re-apply their schema updates, which would break.

I don't think it's worth worrying about in practice though? Could be wrong.